File: fm-dddrive/src/main/java/io/dddrive/core/ddd/model/RepositoryDirectorySPI.java
package io.dddrive.core.ddd.model;

import io.dddrive.core.enums.model.Enumerated;
import io.dddrive.core.enums.model.Enumeration;

public interface RepositoryDirectorySPI {

	void addRepository(Class<? extends Aggregate> intfClass, AggregateRepository<? extends Aggregate> repo);

	<A extends Aggregate> void addPartRepository(Class<? extends Part<A>> intfClass, PartRepository<A, ? extends Part<A>> repo);

	void addPersistenceProvider(Class<? extends Aggregate> intfClass, AggregatePersistenceProvider<? extends Aggregate> repo);

	<P extends Part<?>> void addPartPersistenceProvider(Class<P> intfClass, PartPersistenceProvider<P> ppp);

	<E extends Enumerated> void addEnumeration(Class<E> enumClass, Enumeration<E> enumeration);

}

File: fm-dddrive/src/main/java/io/dddrive/core/ddd/model/RepositoryDirectorySPI.kt
package io.dddrive.core.ddd.model

import io.dddrive.core.enums.model.Enumerated
import io.dddrive.core.enums.model.Enumeration

interface RepositoryDirectorySPI {

	fun addRepository(
		intfClass: Class<out Aggregate>,
		repo: AggregateRepository<out Aggregate>,
	)

	fun <A : Aggregate> addPartRepository(
		intfClass: Class<out Part<A>>,
		repo: PartRepository<A, out Part<A>>,
	)

	fun addPersistenceProvider(
		intfClass: Class<out Aggregate>,
		repo: AggregatePersistenceProvider<out Aggregate>,
	)

	fun <P : Part<*>> addPartPersistenceProvider(
		intfClass: Class<P>,
		ppp: PartPersistenceProvider<P>,
	)

	fun <E : Enumerated> addEnumeration(
		enumClass: Class<E>,
		enumeration: Enumeration<E>,
	)

}
